### PR TITLE
Dockerfile and scripts for setting up docker image

### PR DESCRIPTION
**Pull request by KZhu-ME**
The GPU side of the dockerfile still isn't complete, it errors out, but the CPU docker stuff should all be working fine.

### CODE REVIEW

**Review by nruof — APPROVED** [review]

File: medocker/docker_shared/make_docker.sh [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+if [ ! -d "../../../ME_warp" ]; then
+    echo "No ME_warp install found in ../../.."
+    exit 111
+elif [ ! -d "../../metools" ]; then 
+    echo "No metools install found in ../.."
+    exit 111
+elif [ ! -d "../../../minerva" ]; then 
+    echo "No minerva install found in ../../.."
+    exit 111
+fi
```
Comment (roelof-groenewald) [inline medocker/docker_shared/make_docker.sh]: Why is this needed? We shouldn't need any of those packages, no?

File: medocker/docker_shared/make_docker.sh [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+# Do the docker build
+cp .dockerignore ../../../
+$(aws ecr get-login --no-include-email --region us-west-2)
+docker build --build-arg BUILDPLATFORM=$(uname -m) -f Dockerfile -t simteam/metools-"$FULLTAG" ../../../
+docker tag simteam/metools-"$FULLTAG" 167833485543.dkr.ecr.us-west-2.amazonaws.com/simteam/metools-"$FULLTAG"
+docker push 167833485543.dkr.ecr.us-west-2.amazonaws.com/simteam/metools-"$FULLTAG"
```
Comment (roelof-groenewald) [inline medocker/docker_shared/make_docker.sh]: These `metools` labels should be changed to `mewarpx`.

File: medocker/docker_shared/run_warpx.sh [hunk 3]
```diff
@@ -0,0 +1,133 @@
+#!/bin/bash
+
+# This script handles running MPI/Warp in docker containers on AWS, and the
```
Comment (roelof-groenewald) [inline medocker/docker_shared/run_warpx.sh]:
```suggestion
# This script handles running MPI/WarpX in docker containers on AWS, and the
```

File: medocker/docker_shared/run_warpx.sh [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+fi
+
+# Use EFS directory as working directory.
+echo Make directory ${DIRNAME} in EFS and cd there
+mkdir -p /efs/WarpX_simulation_runs/${DIRNAME}
+cd /efs/WarpX_simulation_runs/${DIRNAME}
+
+# Copy files from S3
+aws s3 cp --recursive s3://${BUCKET}/${DIRNAME} ./ --exclude "*" \
+    --include "run*" --include "*.dump" --include "std*" \
```
Comment (roelof-groenewald) [inline medocker/docker_shared/run_warpx.sh:39]: `.dump` files are Warp specific and isn't (so far) used in WarpX.

File: medocker/docker_shared/make_docker.sh (hunk 1 above)
Comment (KZhu-ME) [reply medocker/docker_shared/make_docker.sh]: Good catch. We tested the docker image manually and didn't use this file, and forgot to update it.

File: medocker/docker_shared/run_warpx.sh [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+# other files (which may still be opened!)
+function checkpoint {
+while true
+do
+    sleep 300
+    if [[ "$MVFILES" == "True" ]]
+    then
+        find . -name "*Header" | while read file; do checkptmove "$file"; done
+        find . -name "Cell_D_*[0-9][0-9][0-9][0-9][0-9]" | while read file; do checkptmove "$file"; done
+        find . -name "warpx_job_info" | while read file; do checkptmove "$file"; done
```
Comment (roelof-groenewald) [inline medocker/docker_shared/run_warpx.sh:77]: We should have a discussion about whether we think copying the full diagnostic outputs to S3 is what we want to do since it will result in a lot of storage usage. Perhaps we only need to copy summarized data.

File: medocker/docker_shared/run_warpx.sh (hunk 4 above)
Comment (KZhu-ME) [reply medocker/docker_shared/run_warpx.sh:39]: I can remove it, but it shouldn't be a problem, since the script just checks to see if any dump files exists and then copies them if they exist. We could keep this line in if we plan on using .dump files for warpx as well.

File: medocker/docker_shared/run_warpx.sh (hunk 5 above)
Comment (KZhu-ME) [reply medocker/docker_shared/run_warpx.sh:77]: Definitely, since we're using elastic storage now. I believe that this script should still be fine as is, because it allows users to choose either to copy the diagnostic files over or move them. To prevent redundant use of space, we could use the move functionality instead of copy.

File: medocker/docker_shared/run_warpx.sh (hunk 4 above)
Comment (roelof-groenewald) [reply medocker/docker_shared/run_warpx.sh:39]: The .dump format is specific to Warp so I don't see a way in which we would start using it but I agree it doesn't hurt anything.

**Comment by roelof-groenewald** [comment]
@KZhu-ME FYI there is a tab versus 4 space issue in some file that is causing the style check to fail.

File: medocker/docker_shared/run_warpx.sh (hunk 5 above)
Comment (peterscherpelz) [reply medocker/docker_shared/run_warpx.sh:77]: @roelof-groenewald I'd be fine with a flag that only optionally copies over full diagnostics. If we do that, we should also write a script that goes through the efs drive and deletes old runs/old diagnostics, because EFS storage is more expensive than S3 by quite a bit.

File: medocker/docker_ubuntu2004/.dockerignore [hunk 6]
```diff
@@ -0,0 +1,4 @@
+runtest/diags
+runtest/results.p
+runtest/results.txt
+runtest/SETUP
```
Comment (peterscherpelz) [inline medocker/docker_ubuntu2004/.dockerignore:4]: I think this file is probably deprecated and can be deleted? I don't remember what it was used for, and looks old judging by the file names.

File: medocker/docker_ubuntu2004/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+# An alternative is to do "docker run --cap-add SYS_PTRACE ...", which keeps CMA.
+# ENV OMPI_MCA_btl_vader_single_copy_mechanism none
+
+# DIRECTORY is used to select S3 directory
+ENV DIRECTORY=null
+# BUCKET is used to select S3 bucket
+ENV BUCKET=null
+
+WORKDIR /merunset/run
+# ENTRYPOINT ["/bin/bash", "--login", "-c", "source /etc/profile && ./run_warpx.sh $DIRECTORY $BUCKET"]
```
Comment (peterscherpelz) [inline medocker/docker_ubuntu2004/Dockerfile]: Why is this commented out? IIRC this made it auto-execute regardless of correctly entering all the info in to the AWS Batch job definition, which was a nice simplification.

File: medocker/docker_shared/python_packages.sh [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+     pykern \
+     yt
+
+${INSTALL_CMD} scikit-image
+
+# This is necessary for S3 copying
+${INSTALL_CMD} awscli
+
+# Install explicitly, rather than as a REQUIRES of metools so that we pick up the installed cmake
+${INSTALL_CMD} "line-profiler==3.2.1"
```
Comment (peterscherpelz) [inline medocker/docker_shared/python_packages.sh]: Either change metools -> mewarpx in comment or delete this; I don't think we're using line_profiler at this point?

File: medocker/docker_shared/README [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+│   │   │   └── run_warp.sh
+│   ├── mewarpx
+├── warpx-data
+
+- Move into the docker_<yourimagefolder>
+- Make sure the aws command line is installed and your credentials are
+  configured.
+- Run ../docker_shared/make_docker.sh <current git branch or tag> <yourimagesuffix>
+  (eg with a suffix '-ubuntumpich', and a git branch 'psdevel', this will push
+   to simteam/metools-psdevel-ubuntumpich)
```
Comment (peterscherpelz) [inline medocker/docker_shared/README]:
```suggestion
To compile and upload:
- Create a directory with this structure:
├── WarpX
│   ├── medocker
│   │   ├── docker_<yourimagefolder>
│   │   │   └── Dockerfile
│   │   ├── docker_shared
│   │   │   ├── make_docker.sh
│   │   │   ├── README
│   │   │   └── run_warpx.sh
│   ├── mewarpx
├── warpx-data

- Move into the docker_<yourimagefolder>
- Make sure the aws command line is installed and your credentials are
  configured.
- Run ../docker_shared/make_docker.sh <current git branch or tag> <yourimagesuffix>
  (eg with a suffix '-ubuntumpich', and a git branch 'psdevel', this will push
   to simteam/mewarpx-psdevel-ubuntumpich)
```

**Review by peterscherpelz — CHANGES_REQUESTED** [review]

File: medocker/docker_ubuntu2004/Dockerfile (hunk 7 above)
Comment (KZhu-ME) [reply medocker/docker_ubuntu2004/Dockerfile]: I think it was commented out when we were testing the dockerfile, and I forgot to uncomment it.

File: medocker/docker_shared/ubuntu_packages.sh [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+    gfortran            \
+    git                 \
+    gnupg               \
+    llvm-9-dev          \
+    libblosc-dev        \
+    libffi-dev          \
+    libfreetype6-dev    \
+    libhdf5-dev         \
+    liblapack-dev       \
+    liblapacke-dev      \
```
Comment (EtherealMC-Bit) [inline medocker/docker_shared/ubuntu_packages.sh:27]: Just want to make sure that `liblapack-dev` and `liblapacke-dev` are separate packages and not a typo or something.

**Review by EtherealMC-Bit — APPROVED** [review]
LGTM with one question

**Review by peterscherpelz — APPROVED** [review]

File: medocker/docker_shared/ubuntu_packages.sh (hunk 10 above)
Comment (peterscherpelz) [reply medocker/docker_shared/ubuntu_packages.sh:27]: Looks like, see https://packages.debian.org/sid/liblapacke-dev
